src/runtime/runtime.js

```diff
@@ -96,291 +83,418 @@
-   */
-  function isShimSymbol(symbol) {
-    return typeof symbol === 'object' && symbol instanceof SymbolValue;
+  var CONTINUATION_TYPE = Object.create(null);
+
+  function createContinuation(operand, thisArg, argsArray) {
+    return [CONTINUATION_TYPE, operand, thisArg, argsArray];
   }
 
-  function typeOf(v) {
-    if (isShimSymbol(v))
-      return 'symbol';
-    return typeof v;
+  function isContinuation(object) {
+    return object && object[0] === CONTINUATION_TYPE;
   }
 
-  /**
-   * Creates a new unique symbol object.
-   * @param {string=} string Optional string used for toString.
-   * @constructor
-   */
-  function Symbol(description) {
-    var value = new SymbolValue(description);
-    if (!(this instanceof Symbol))
-      return value;
+  var isTailRecursiveName = createPrivateName();
 
-    // new Symbol should throw.
-    //
-    // There are two ways to get a wrapper to a symbol. Either by doing
-    // Object(symbol) or call a non strict function using a symbol value as
-    // this. To correctly handle these two would require a lot of work for very
-    // little gain so we are not doing those at the moment.
-    throw new TypeError('Symbol cannot be new\'ed');
+  function initTailRecursiveFunction(func) {
+    func[isTailRecursiveName] = true;
+    return func;
   }
 
-  $defineProperty(Symbol.prototype, 'constructor', nonEnum(Symbol));
-  $defineProperty(Symbol.prototype, 'toString', method(function() {
-    var symbolValue = this[symbolDataProperty];
-    return symbolValue[symbolInternalProperty];
-    /* The implementation of toString below matches the spec, but prevents
-    use of Symbols in eg generators unless --symbol is set. To simplify our
-    code we deliberately go against the spec here.
-    if (!symbolValue)
-      throw TypeError('Conversion from symbol to string');
-    var desc = symbolValue[symbolDescriptionProperty];
-    if (desc === undefined)
-      desc = '';
-    return 'Symbol(' + desc + ')';
-    */
-  }));
-  $defineProperty(Symbol.prototype, 'valueOf', method(function() {
-    var symbolValue = this[symbolDataProperty];
-    if (!symbolValue)
-      throw TypeError('Conversion from symbol to string');
-    if (!getOption('symbols'))
-      return symbolValue[symbolInternalProperty];
-    return symbolValue;
-  }));
-
-  function SymbolValue(description) {
-    var key = newUniqueString();
-    $defineProperty(this, symbolDataProperty, {value: this});
-    $defineProperty(this, symbolInternalProperty, {value: key});
-    $defineProperty(this, symbolDescriptionProperty, {value: description});
-    freeze(this);
-    symbolValues[key] = this;
+  function isTailRecursive(func) {
+    return func[isTailRecursiveName] || false;
   }
-  $defineProperty(SymbolValue.prototype, 'constructor', nonEnum(Symbol));
-  $defineProperty(SymbolValue.prototype, 'toString', {
-    value: Symbol.prototype.toString,
-    enumerable: false
+
+  function tailCall(func, thisArg, argArray) {
```

Comment:
I had had this in an extra module `tailCalls.js` until I wanted to compile the whole Traceur codebase with PTC support. Then I ran into the issue that the transformed runtime itself needed bits of `tailCalls.js`, so I had to move it at the top. Because `tailCalls.js` needed `$traceurRuntime` to add properties, I merged it into the main runtime file.

A probably better solution (which would add more globals, though) would be to change `$traceurRuntime.tailcall` to `$traceurRuntime$tailCall`, etc., so that the PTC part of the runtime can be loaded at the very beginning independent of `$traceurRuntime` already set up.

As every second call in a PTC-transpiled program will be a call to `$traceurRuntime.call`, it may even speed things up if one removes the extra property lookup on `$traceurRuntime` by going global.

What do you think?